### PR TITLE
rafthttp: mark unreachable on unexpected response

### LINKED ISSUE #3652: mark endpoint unreachable when endpoint returns unexpected response
In rafthttp, when making request to some endpoint, it may receive response with unexpected status code. This indicates the endpoint doesn't function correctly. It should print warning to notify users about this, and mark it unreachable.

from https://github.com/coreos/etcd/pull/3648/files#r41349972

### PR DESCRIPTION
**Pull request by yichengq**
In rafthttp, when making request to some endpoint, it may receive
response with unexpected status code and header. This indicates the endpoint
doesn't function correctly. It should mark the endpoint unreachable.

Also make minimal changes to make code much readable.

fixes #3652

### CODE REVIEW

File: rafthttp/pipeline.go [hunk 1]
```diff
@@ -153,14 +153,18 @@ func (p *pipeline) post(data []byte) (err error) {
...
-	if err == errMemberRemoved {
-		reportCriticalError(err, p.errorc)
-		return nil
+	if err != nil {
+		p.picker.unreachable(u)
+		// errMemberRemoved is a critical error since a removed member should
+		// always be stopped. So we use reportCriticalError to report it to errorc.
+		if err == errMemberRemoved {
+			reportCriticalError(err, p.errorc)
+			return nil
```
Comment (xiang90) [inline rafthttp/pipeline.go]: please explain why do we sallow the error here?

File: rafthttp/stream.go [hunk 2]
```diff
@@ -392,45 +392,52 @@ func (cr *streamReader) dial(t streamType) (io.ReadCloser, error) {
 	lv := semver.Must(semver.NewVersion(version.Version))
 	if compareMajorMinorVersion(rv, lv) == -1 && !checkStreamSupport(rv, t) {
 		resp.Body.Close()
+		cr.picker.unreachable(u)
```
Comment (xiang90) [inline rafthttp/stream.go]: i do not think we should mark this as unreachable. this is reachable, but the version is not compatible?

File: rafthttp/stream.go [hunk 3]
```diff
@@ -392,45 +392,52 @@ func (cr *streamReader) dial(t streamType) (io.ReadCloser, error) {
...
 		resp.Body.Close()
-		err := fmt.Errorf("the member has been permanently removed from the cluster")
-		select {
-		case cr.errorc <- err:
-		default:
+		cr.picker.unreachable(u)
+		// errMemberRemoved is a critical error since a removed member should
+		// always be stopped. So we use reportCriticalError to report it to errorc.
+		if err == errMemberRemoved {
+			reportCriticalError(err, cr.errorc)
```
Comment (xiang90) [inline rafthttp/stream.go]: why not return nil in this case as before?

File: rafthttp/stream.go [hunk 4]
```diff
@@ -392,45 +392,52 @@ func (cr *streamReader) dial(t streamType) (io.ReadCloser, error) {
...
 		}
 		return nil, err
+	}
+
+	return resp.Body, nil
+}
+
+func (cr *streamReader) checkResponse(resp *http.Response) error {
```
Comment (xiang90) [inline rafthttp/stream.go]: checkResponseHTTPStatus. let us be more clear about func name.

Also can we add a few comments? there are too many cases in side this function.

File: rafthttp/pipeline.go (hunk 1 above)
Comment (yichengq) [inline rafthttp/pipeline.go]: I think it should return error to be more readable. I didn't change it in this PR for clarity. As we come here, I'd like to send a separate commit to address it.

File: rafthttp/stream.go (hunk 2 above)
Comment (yichengq) [inline rafthttp/stream.go]: Based on our previous discussion, as long as it doesn't function as expectation, it is unreachable in the view of local etcd. So I report `unreachable` to the picker.

File: rafthttp/stream.go (hunk 4 above)
Comment (yichengq) [inline rafthttp/stream.go]: The function may also read the body of the response, so `HTTPStatus` is not accurate. `checkHTTPResponse` could be a better naming. I will add comment for this function to describe it.

> Also can we add a few comments? 

sure

File: rafthttp/stream.go (hunk 2 above)
Comment (xiang90) [inline rafthttp/stream.go]: ok.

File: rafthttp/stream.go (hunk 3 above)
Comment (yichengq) [inline rafthttp/stream.go]: the previous case is changed to return err.

**Comment by yichengq** [comment]
Addressed all comments.

File: rafthttp/stream.go [hunk 5]
```diff
@@ -392,45 +392,61 @@ func (cr *streamReader) dial(t streamType) (io.ReadCloser, error) {
...
 		return nil, err
+	}
+
+	return resp.Body, nil
+}
+
+// checkStreamResponse checks the response of one stream request.
+// It returns nil if the response indicates that stream request succeeds,
+// or it returns error.
+// If the response indicates stream request failure, the function may read the
```
Comment (xiang90) [inline rafthttp/stream.go]: I would suggest to make this behavior consistent. If there is an error, always read the body and close it?

File: rafthttp/stream.go [hunk 6]
```diff
@@ -392,45 +392,61 @@ func (cr *streamReader) dial(t streamType) (io.ReadCloser, error) {
...
+// response body for verbosal failure message.
+func (cr *streamReader) checkStreamResponse(resp *http.Response) error {
```
Comment (xiang90) [inline rafthttp/stream.go]: This is ugly. We should not repeat `Stream` since this is already a streamReader. I think I have suggested something else.

Comment (xiang90) [inline rafthttp/stream.go]: And you are checking the HTTPStatusCode of the response. Not the whole response.

File: rafthttp/stream.go (hunk 5 above)
Comment (yichengq) [inline rafthttp/stream.go]: ok

File: rafthttp/stream.go (hunk 6 above)
Comment (yichengq) [inline rafthttp/stream.go]: ok

**Comment by yichengq** [comment]
Addressed all the comments. ready for the review.

File: rafthttp/stream.go [hunk 7]
```diff
@@ -392,45 +392,63 @@ func (cr *streamReader) dial(t streamType) (io.ReadCloser, error) {
...
+// It returns nil if the response indicates that stream request succeeds,
+// or it returns error.
+// If there is an error, the function always reads the response body
+// and closes it.
+func (cr *streamReader) checkResponseStatusCode(resp *http.Response) error {
+	if resp.StatusCode == http.StatusOK {
+		return nil
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
```
Comment (xiang90) [inline rafthttp/stream.go]: I found there is another function called `checkPostResponse`. They share some logic actually. If you want to clean things up, can you please give it a look? Try to merge the logic if possible?

Comment (yichengq) [inline rafthttp/stream.go]: The HTTP status that returns by stream and post are different in some cases. And I don't want to mess them together. For example, for post request, http.StatusNoContent means it succeeds, but if stream request receives this, it is an unexpected case.

Comment (xiang90) [inline rafthttp/stream.go]: > The HTTP status that returns by stream and post are different in some cases. 

You are wrong. They have exact same logic to handle `http.StatusPreconditionFailed` and `http.StatusForbidden`. Please do not repeat them here. Also please make the names of the two checking function to be consistent and put them into the consistent file. (not one in util and one in the pipeline)

Comment (yichengq) [inline rafthttp/stream.go]: `http.StatusForbidden` doesn't share the same logic, and `http.StatusPreconditionFailed` shares the same logic.

I would like to use follow-up PR to make them consistent. This PR only focuses on `mark unreachable` correctly, and I makes minimal changes to make it much readable.

Comment (xiang90) [inline rafthttp/stream.go]: > I would like to use follow-up PR to make them consistent. This PR only focuses on mark unreachable correctly, and I makes minimal changes to make it much readable.

This is fine. Then please keep it minimal and remove all the refactoring code.

Comment (xiang90) [inline rafthttp/stream.go]: `http.StatusForbidden` I have no idea why they do not. Probably due to some bad implementation.

**Comment by yichengq** [comment]
Remove the refactoring code. I will clean up code together in separate clean-up PRs.

**Comment by xiang90** [comment]
LGTM

**Pull request merged**